docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+
+The above list is not comprehensive but a general pattern emerges. The outages range from:
+1. Shutdown of machines in a specific availability zone due to infrastrucure failure which in turn could be due to many reasons listed above.
+2. Network connectivity to the machines running in an availability zone is either severly impacted or broken.
+3. Subset of essential services(e.g. EBS volumes in case of AWS provider) are unhealthy which might also have a cascading effect on other services.
+
+One can further derive that either there is a complete breakdown of an entire availability zone (1 and 2 above) or there is a degradation or unavailability of a subset of essential services.
+In the first version of this document we define a AZ outage only when either of (1) or (2) occurs as defined above.
+
+#### Impact of a Zone Outage
+
+As part of the current recovery mechanisms (described above), if `Machine-Controller-Manager` is able to delete the machines then per `MachineDeployment` it will delete one machine at a time and wait for a new machine to transition from `Pending` to `Running` state. In case of a network unreachability it will be able to delete a machine and subsequently launch a new machine but the newly launched machine will be stuck in `Pending` state as the `Kubelet` running in the machine will not be able to create its lease. There will also not be any corresponding `Node` object for the newly launched machine. Rest of the machines in this `MachineDeployment` will be stuck in `Unknown` state.
+
+**Kube API Server & Gardener Resource Manager**
+
+These pods are stateless, losing one pod can be tolerated since there will be two other replicas that will continue to run in other two zones which are available (considering that there are 3 zones in a region). Pod Anti-Affinity rule is set with `preferredDuringSchedulingIgnoredDuringExecution` allowing rescheduling of a pod to another zone and any available node which does not have a running Kube API Server/Gardener Resource Manager pod. (See current recovery mechanisms described above to see when these pods will be rescheduled).
+
+**etcd**
+
+A minimum and default size of HA etcd cluster setup is 3. This allows tolerance of one AZ failure. If more than one AZ fails or is unreachable then etcd cluster will lose its quorum. Pod Anti-Affinity policies that are initially set will not allow automatic rescheduling of etcd pod onto another zone (unless ofcourse affinity rules are dynamically changed). Reference to `PVC` will also prevent an `etcd` member from getting scheduled in another zone since persistent volumes are not shared across availability zones. 
+
+**Other Shoot Control Plane Components**
+
+All the other shoot control plane components have:
+* Single replicas
+* No affinity rules influecing their scheduling
+
```

Comment:
See above, strictly speaking `gardenlet` and `istio-ingressgateway` are no "control plane" components, but they are essential seed components that should be probably replicated.